### DISCUSSION TITLE
Accessing deeply nested properties of Optional

### DISCUSSION THREAD

**Discussion by kohenkatz** [spatie/laravel-data#493] [Q&A]
I have LaravelData classes that look something like this:

```
class A extends Data
{
    public function __construct(
        public B|Optional $b,
        public C|Optional $c,
    ) {
    }
}

class C extends Data
{
    public function __construct(
        public string $str,
        public D|Optional $d,
    ) {
    }
}

class D extends Data
{
    public function __construct(
        public string|Optional $str,
    ) {
    }
}
```

With an instance of `A` being a field on an Eloquent model, I want to be able to access something like `$myObj->a->c->str` in the simplest way possible, but also easily show a default value if the value is not set.

If I used nullable (e.g. `public ?C $c` instead of `C|Optional`), I would be able to use a nullsafe operator like this: `$myObj->a?->c?->str ?: 'default`. However, using nullable instead of Optional means that unset properties are written as explicit nulls in the database, which is not what I want.

I don't want to be stuck writing lots of `if`s (or nasty nested ternaries) like this:

```
$str = 'default';
if (!$myObj->a instanceof Optional) {
    if (!$myObj->a->c instanceof Optional) {
        // etc...
    }
}
```

**Comment by rubenvanassche** [comment]
I don't think there's an easier way other than using if's or using nulls since we don't have a thing like operator overloading in PHP.

**Reply by bastianschwarz** [reply]
Just stumbled across this. Would you be open for something like

```php
<?php

namespace Spatie\LaravelData;

use RuntimeException;

class Optional
{
    public function __get(string $name)
    {
        return null;
    }

    public function __set(string $name, $value): void
    {
        throw new RuntimeException('Setting properties on OptionalNull is not allowed');
    }

    public function __isset(string $name): bool {
        return false;
    }

    public static function create(): Optional
    {
        return new self();
    }
}
```

Sure, magic property access is not the greatest thing but it would enable the "mostly" usual null operations in php

`$a->$b->$c`

I think this could be made fancy when the Optional is aware of the possible types so isset could check if the property could exist but this might be a start.

Thoughts?

EDIT: I tried creating my own subtype of Optional but the creation is hardcoded to the Optional class so that didn't work out. I tend to use the Data objects as Request without transforming so I can have type checks and autocomplete for the properties.
